mongodb_exporter.go

```diff
@@ -11,137 +12,156 @@ import (
+	sslKeyFileF    = flag.String("web.ssl-key-file", "", "Path to SSL key file.")
+
+	uriF     = flag.String("mongodb.uri", mongodbDefaultURI(), "MongoDB URI, format: [mongodb://][user:pass@]host1[:port1][,host2[:port2],...][/database][?options]")
+	tlsF     = flag.Bool("mongodb.tls", false, "Enable tls connection with mongo server")
+	tlsCertF = flag.String("mongodb.tls-cert", "", "Path to PEM file that contains the certificate (and optionally also the private key in PEM format).\n"+
 		"    \tThis should include the whole certificate chain.\n"+
 		"    \tIf provided: The connection will be opened via TLS to the MongoDB server.")
-	mongodbTlsPrivateKey = flag.String("mongodb.tls-private-key", "", "Path to PEM file that conains the private key (if not contained in mongodb.tls-cert file).")
-	mongodbTlsCa         = flag.String("mongodb.tls-ca", "", "Path to PEM file that conains the CAs that are trused for server connections.\n"+
+	tlsPrivateKeyF = flag.String("mongodb.tls-private-key", "", "Path to PEM file that contains the private key (if not contained in mongodb.tls-cert file).")
+	tlsCAF         = flag.String("mongodb.tls-ca", "", "Path to PEM file that contains the CAs that are trusted for server connections.\n"+
 		"    \tIf provided: MongoDB servers connecting to should present a certificate signed by one of this CAs.\n"+
 		"    \tIf not provided: System default CAs are used.")
-	mongodbTlsDisableHostnameValidation = flag.Bool("mongodb.tls-disable-hostname-validation", false, "Do hostname validation for server connection.")
+	tlsDisableHostnameValidationF = flag.Bool("mongodb.tls-disable-hostname-validation", false, "Do hostname validation for server connection.")
+
+	enabledGroupsFlag = flag.String("groups.enabled", "asserts,durability,background_flushing,connections,extra_info,global_lock,index_counters,network,op_counters,op_counters_repl,memory,locks,metrics", "Comma-separated list of groups to use, for more info see: docs.mongodb.org/manual/reference/command/serverStatus/")
 )
 
 var landingPage = []byte(`<html>
 <head><title>MongoDB exporter</title></head>
 <body>
 <h1>MongoDB exporter</h1>
-<p><a href='` + *metricsPathFlag + `'>Metrics</a></p>
+<p><a href='` + *metricsPathF + `'>Metrics</a></p>
 </body>
 </html>
 `)
 
-func printVersion() {
-	fmt.Printf("mongodb_exporter version: %s, git commit hash: %s\n", version, versionGitCommit)
-}
-
 type webAuth struct {
-	User     string `yaml:"server_user,omitempty"`
+	Username string `yaml:"server_user,omitempty"`
 	Password string `yaml:"server_password,omitempty"`
 }
 
 type basicAuthHandler struct {
-	handler  http.HandlerFunc
-	user     string
-	password string
+	webAuth
+	handler http.HandlerFunc
 }
 
 func (h *basicAuthHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	user, password, ok := r.BasicAuth()
-	if !ok || password != h.password || user != h.user {
-		w.Header().Set("WWW-Authenticate", "Basic realm=\"metrics\"")
+	username, password, _ := r.BasicAuth()
```

Comment:
If not ok, `username` and `password` will be empty strings which are good enough for us.